File: src/tauri/src/commands/storage.rs
//! Storage UI: enumerate and clear `~/.tomat/` content.
//!
//! The tree builders walk directories synchronously because issuing a
//! tokio task per file is heavier than the blocking read cost on the
//! typical working set; this keeps the storage view snappy without async
//! overhead on what is effectively a stat-heavy traversal.

use crate::commands::paths::resolve_within;
use crate::commands::session::{read_session_file_sync, SessionFile};
use crate::commands::settings::clear_secrets;
use crate::commands::snippets::Snippet;
use crate::error::{AppError, AppResult};
use std::path::{Path, PathBuf};
use tauri::{AppHandle, Manager};

#[derive(serde::Serialize)]
#[serde(tag = "kind", rename_all = "lowercase")]
pub enum StorageNode {
    File {
        name: String,
        path: String,
        size: u64,
    },
    Folder {
        name: String,
        path: String,
        size: u64,
        children: Vec<StorageNode>,
    },
}

#[derive(serde::Serialize)]
pub struct StorageTree {
    pub models: Vec<StorageNode>,
    pub sessions: Vec<StorageNode>,
    pub snippets: Vec<StorageNode>,
    pub total_size: u64,
    pub models_size: u64,
    pub sessions_size: u64,
    pub snippets_size: u64,
    pub settings_size: u64,
    pub root_path: String,
}

fn dir_size_recursive(path: &Path) -> u64 {
    let mut total: u64 = 0;
    let Ok(entries) = std::fs::read_dir(path) else {
        return 0;
    };
    for entry in entries.flatten() {
        let p = entry.path();
        if let Ok(meta) = entry.metadata() {
            if meta.is_file() {
                total += meta.len();
            } else if meta.is_dir() {
                total += dir_size_recursive(&p);
            }
        }
    }
    total
}

fn storage_name(node: &StorageNode) -> &str {
    match node {
        StorageNode::File { name, .. } => name.as_str(),
        StorageNode::Folder { name, .. } => name.as_str(),
    }
}

fn storage_size(node: &StorageNode) -> u64 {
    match node {
        StorageNode::File { size, .. } => *size,
        StorageNode::Folder { size, .. } => *size,
    }
}

// File extensions that count as "model / support files" we want to surface
// in the storage UI. Includes gguf (llama.cpp), bin (whisper.cpp + some ORT
// voice tensors), onnx + json (Kokoro / transformers.js), and pt (PyTorch
// weight tensors, used by Kokoro-style voice files).
const MODEL_FILE_EXTS: &[&str] = &["gguf", "bin", "onnx", "json", "pt"];

fn collect_model_files(base: &Path, dir: &Path, out: &mut Vec<StorageNode>) {
    let Ok(entries) = std::fs::read_dir(dir) else {
        return;
    };
    for entry in entries.flatten() {
        let p = entry.path();
        let Ok(meta) = entry.metadata() else { continue };
        if meta.is_dir() {
            collect_model_files(base, &p, out);
        } else if meta.is_file() {
            let ext = p
                .extension()
                .and_then(|e| e.to_str())
                .map(str::to_ascii_lowercase)
                .unwrap_or_default();
            if !MODEL_FILE_EXTS.contains(&ext.as_str()) {
                continue;
            }
            let rel = p.strip_prefix(base).unwrap_or(&p);
            let name = rel.to_string_lossy().to_string();
            out.push(StorageNode::File {
                name,
                path: p.to_string_lossy().to_string(),
                size: meta.len(),
            });
        }
    }
}

fn build_models_tree(models_dir: &Path) -> Vec<StorageNode> {
    let mut out = Vec::new();
    let Ok(user_entries) = std::fs::read_dir(models_dir) else {
        return out;
    };

    for user_entry in user_entries.flatten() {
        let user_path = user_entry.path();
        if !user_path.is_dir() {
            continue;
        }
        let Ok(repo_entries) = std::fs::read_dir(&user_path) else {
            continue;
        };
        for repo_entry in repo_entries.flatten() {
            let repo_path = repo_entry.path();
            if !repo_path.is_dir() {
                continue;
            }
            let user_name = user_entry.file_name().to_string_lossy().to_string();
            let repo_name = repo_entry.file_name().to_string_lossy().to_string();
            let display_name = format!("{user_name}/{repo_name}");

            let mut files = Vec::new();
            collect_model_files(&repo_path, &repo_path, &mut files);
            files.sort_by(|a, b| storage_name(a).cmp(storage_name(b)));

            let has_mmproj = files.iter().any(|n| {
                let name = storage_name(n);
                let basename = name.rsplit('/').next().unwrap_or(name);
                basename.to_ascii_lowercase().starts_with("mmproj")
            });

            if files.is_empty() {
                continue;
            }

            if has_mmproj || files.len() > 1 {
                let size: u64 = files.iter().map(storage_size).sum();
                out.push(StorageNode::Folder {
                    name: display_name,
                    path: repo_path.to_string_lossy().to_string(),
                    size,
                    children: files,
                });
            } else {
                // Single-file repo: render inline "user/repo/filename" so the
                // list stays dense when only one file matters (e.g. whisper).
                let Some(file) = files.into_iter().next() else {
                    continue;
                };
                if let StorageNode::File {
                    name: fname,
                    path,
                    size,
                } = file
                {
                    out.push(StorageNode::File {
                        name: format!("{display_name}/{fname}"),
                        path,
                        size,
                    });
                }
            }
        }
    }
    out.sort_by(|a, b| storage_name(a).cmp(storage_name(b)));
    out
}

fn build_sessions_tree(sessions_dir: &Path) -> Vec<StorageNode> {
    let mut out = Vec::new();
    let Ok(entries) = std::fs::read_dir(sessions_dir) else {
        return out;
    };
    for entry in entries.flatten() {
        let p = entry.path();
        let Ok(meta) = entry.metadata() else { continue };
        if !meta.is_dir() {
            continue;
        }
        let stem = p
            .file_name()
            .and_then(|s| s.to_str())
            .unwrap_or_default()
            .to_string();
        let messages_path = p.join("messages.json");
        if !messages_path.exists() {
            continue;
        }
        let title = read_session_file_sync(&messages_path)
            .map(|s: SessionFile| s.title)
            .unwrap_or_default();
        let display = if title.trim().is_empty() { stem } else { title };
        out.push(StorageNode::File {
            name: display,
            path: p.to_string_lossy().to_string(),
            size: dir_size_recursive(&p),
        });
    }
    out.sort_by(|a, b| storage_name(a).cmp(storage_name(b)));
    out
}

fn build_snippets_tree(snippets_dir: &Path) -> Vec<StorageNode> {
    let mut out = Vec::new();
    let Ok(entries) = std::fs::read_dir(snippets_dir) else {
        return out;
    };
    for entry in entries.flatten() {
        let p = entry.path();
        let Ok(meta) = entry.metadata() else { continue };
        if !meta.is_file() {
            continue;
        }
        if p.extension().and_then(|e| e.to_str()) != Some("json") {
            continue;
        }
        let stem = p
            .file_stem()
            .and_then(|s| s.to_str())
            .unwrap_or_default()
            .to_string();
        // Prefer the snippet's user-facing `name` field when readable; otherwise
        // fall back to the filename stem.
        let display = std::fs::read_to_string(&p)
            .ok()
            .and_then(|c| serde_json::from_str::<Snippet>(&c).ok())
            .map(|s| s.name)
            .filter(|n| !n.trim().is_empty())
            .unwrap_or(stem);
        out.push(StorageNode::File {
            name: display,
            path: p.to_string_lossy().to_string(),
            size: meta.len(),
        });
    }
    out.sort_by(|a, b| storage_name(a).cmp(storage_name(b)));
    out
}

/// Enumerate everything under `~/.tomat/` for the storage UI.
#[tauri::command]
pub async fn list_tomat_storage(handle: AppHandle) -> AppResult<StorageTree> {
    let home = handle.path().home_dir()?;
    let root = home.join(".tomat");
    let models_dir = root.join("models");
    let sessions_dir = root.join("sessions");
    let snippets_dir = root.join("snippets");

    let models = build_models_tree(&models_dir);
    let sessions = build_sessions_tree(&sessions_dir);
    let snippets = build_snippets_tree(&snippets_dir);

    let models_size: u64 = models.iter().map(storage_size).sum();
    let sessions_size: u64 = sessions.iter().map(storage_size).sum();
    let snippets_size: u64 = snippets.iter().map(storage_size).sum();
    let settings_size = std::fs::metadata(root.join("settings.json"))
        .map(|m| m.len())
        .unwrap_or(0);
    let total_size = dir_size_recursive(&root);

    Ok(StorageTree {
        models,
        sessions,
        snippets,
        total_size,
        models_size,
        sessions_size,
        snippets_size,
        settings_size,
        root_path: root.to_string_lossy().to_string(),
    })
}

/// Delete the given paths. Every path is canonicalized and must resolve under `~/.tomat/`.
#[tauri::command]
pub async fn delete_tomat_paths(handle: AppHandle, paths: Vec<String>) -> AppResult<()> {
    let home = handle.path().home_dir()?;
    let root_canonical = home.join(".tomat").canonicalize()?;

    for path_str in paths {
        let p = PathBuf::from(&path_str);
        let canonical = resolve_within(&p, &root_canonical)
            .ok_or_else(|| AppError::validation(format!("Path not under ~/.tomat: {path_str}")))?;
        let meta = tokio::fs::metadata(&canonical).await?;
        if meta.is_dir() {
            tokio::fs::remove_dir_all(&canonical).await?;
        } else {
            tokio::fs::remove_file(&canonical).await?;
        }
    }
    Ok(())
}

/// Remove all downloaded models from `~/.tomat/models/`.
#[tauri::command]
pub async fn clear_tomat_models(handle: AppHandle) -> AppResult<()> {
    let home = handle.path().home_dir()?;
    let dir = home.join(".tomat").join("models");
    if dir.exists() {
        tokio::fs::remove_dir_all(&dir).await?;
    }
    tokio::fs::create_dir_all(&dir).await?;
    Ok(())
}

/// Remove every saved session from `~/.tomat/sessions/`.
#[tauri::command]
pub async fn clear_tomat_sessions(handle: AppHandle) -> AppResult<()> {
    let home = handle.path().home_dir()?;
    let dir = home.join(".tomat").join("sessions");
    if dir.exists() {
        tokio::fs::remove_dir_all(&dir).await?;
    }
    tokio::fs::create_dir_all(&dir).await?;
    Ok(())
}

/// Wipe the settings file and every keychain entry (release) or the
/// `.secrets.json` fallback (dev) for the caller-named secrets.
#[tauri::command]
pub async fn clear_tomat_settings(handle: AppHandle, secret_keys: Vec<String>) -> AppResult<()> {
    let home = handle.path().home_dir()?;
    let path = home.join(".tomat").join("settings.json");
    if path.exists() {
        tokio::fs::remove_file(&path).await?;
    }
    clear_secrets(&handle, &secret_keys).await?;
    Ok(())
}

File: packages/tomat-core/src/services/storage.ts
// Storage view for the Settings "Storage" field: a full per-category view of
// everything tomat keeps on disk (downloaded models, sidecar binaries, the
// sessions JSON store, installed toolkits, caches, logs, settings) plus the
// clear/delete operations for each. The core is the sole authority on what may
// be removed: every node is tagged with a `lock_reason` when it's in use, and a
// delete request re-derives that lock server-side and refuses anything locked
// (never trusting the client's view).
//
// In-use ("locked") rules per category:
//   - models:   required by the current settings (requiredModelRefs)
//   - binaries: required by settings OR its sidecar is running
//   - cache:    never (regenerable)
//   - logs:     the live core.log only
//   - sessions: an agent is mid-turn on it (chat activeSessionIds)
//   - toolkits: it has an enabled tool, or it's the built-in toolkit
//   - settings: never; "clearing" it is a factory reset (defaults + wipe secrets)

import { basename, dirname, isAbsolute, join, normalize, relative } from "@std/path";
import type { BinaryKind, StorageCategory, StorageCategoryId, StorageNode } from "@tomat/shared";
import { requiredBinaryKinds, requiredModelRefs } from "@tomat/shared";
import { paths, sessionDir, toolkitDir } from "../paths.ts";
import { AppError } from "../shared/errors.ts";
import { getLogger } from "../shared/log.ts";
import { resolveHfPath } from "../models/manager.ts";
import { binariesManager } from "../binaries/manager.ts";
import { sidecarManager } from "../sidecars/manager.ts";
import { toolkitsRegistry } from "../toolkits/registry.ts";
import { uninstallToolkit } from "../toolkits/uninstall.ts";
import { sessionsRepo } from "./sessions-store.ts";
import { chatService } from "./chat.ts";
import { loadCoreSettingsResolved, resetCoreSettings } from "./core-settings.ts";
import { clearAllSecrets } from "./secrets.ts";
import { notifyRequirementsChanged } from "./requirements.ts";

const log = getLogger("storage");

// Extensions that count as model / support files: gguf (llama.cpp), bin
// (whisper.cpp + some ORT tensors), onnx + json (Kokoro / transformers.js),
// pt (PyTorch voice tensors).
const MODEL_FILE_EXTS = new Set(["gguf", "bin", "onnx", "json", "pt"]);

// Running sidecar kind -> the binary it runs. deno is always required (it runs
// the tts / tool / embedding workers), so it's covered by requiredBinaryKinds.
const SIDECAR_BINARY: Partial<Record<string, BinaryKind>> = {
  llama: "llama-server",
  whisper: "whisper-server",
};

// --- lock context ----------------------------------------------------------

interface LockContext {
  requiredModelPaths: Set<string>;
  lockedBinaryKinds: Set<BinaryKind>;
  activeSessionIds: Set<string>;
  enabledToolkitIds: Set<string>;
}

async function buildLockContext(): Promise<LockContext> {
  const settings = await loadCoreSettingsResolved();

  const requiredModelPaths = new Set(
    requiredModelRefs(settings).map((r) => normalize(resolveHfPath(r.source))),
  );

  const lockedBinaryKinds = new Set<BinaryKind>(requiredBinaryKinds(settings));
  for (const s of sidecarManager().getStatuses()) {
    if (s.status !== "Running") continue;
    const bin = SIDECAR_BINARY[s.kind];
    if (bin) lockedBinaryKinds.add(bin);
  }

  const enabledToolkitIds = new Set<string>();
  try {
    for (const tk of toolkitsRegistry().list()) {
      if (
        toolkitsRegistry()
          .listTools(tk.id)
          .some((t) => t.enabled)
      ) {
        enabledToolkitIds.add(tk.id);
      }
    }
  } catch (err) {
    // A toolkit-registry read failure (e.g. a stale DB schema) must not blank
    // the whole storage view; fall back to "nothing enabled" (the toolkits
    // category itself degrades to empty below).
    log.warn(`toolkit lock scan failed: ${err instanceof Error ? err.message : err}`);
  }

  return {
    requiredModelPaths,
    lockedBinaryKinds,
    activeSessionIds: chatService().activeSessionIds(),
    enabledToolkitIds,
  };
}

// --- node helpers ----------------------------------------------------------

function fileNode(name: string, path: string, size: number, lock?: string): StorageNode {
  return lock
    ? { kind: "file", name, path, size, lock_reason: lock }
    : { kind: "file", name, path, size };
}

function folderNode(
  name: string,
  path: string,
  size: number,
  children: StorageNode[] = [],
  lock?: string,
): StorageNode {
  return lock
    ? { kind: "folder", name, path, size, children, lock_reason: lock }
    : { kind: "folder", name, path, size, children };
}

async function statSize(path: string): Promise<number> {
  try {
    return (await Deno.stat(path)).size;
  } catch {
    return 0;
  }
}

async function dirSize(path: string): Promise<number> {
  let total = 0;
  let entries: AsyncIterable<Deno.DirEntry>;
  try {
    entries = Deno.readDir(path);
  } catch {
    return 0;
  }
  for await (const e of entries) {
    const full = join(path, e.name);
    if (e.isDirectory) total += await dirSize(full);
    else if (e.isFile) total += await statSize(full);
  }
  return total;
}

// --- models ----------------------------------------------------------------

async function collectModelFiles(base: string, dir: string, out: StorageNode[]): Promise<void> {
  let entries: AsyncIterable<Deno.DirEntry>;
  try {
    entries = Deno.readDir(dir);
  } catch {
    return;
  }
  for await (const e of entries) {
    const p = join(dir, e.name);
    if (e.isDirectory) {
      await collectModelFiles(base, p, out);
    } else if (e.isFile) {
      const ext = (e.name.includes(".") ? e.name.split(".").pop()! : "").toLowerCase();
      if (!MODEL_FILE_EXTS.has(ext)) continue;
      const size = await statSize(p);
      out.push({ kind: "file", name: relative(base, p).replaceAll("\\", "/"), path: p, size });
    }
  }
}

function applyModelLock(node: StorageNode, ctx: LockContext): boolean {
  // Annotate a model file/folder with a lock reason in place. Returns whether
  // the node ended up locked (a folder is locked iff every child is).
  if (node.kind === "file") {
    if (ctx.requiredModelPaths.has(normalize(node.path))) {
      node.lock_reason = "Required by current settings";
      return true;
    }
    return false;
  }
  if (node.children.length === 0) return false;
  let allLocked = true;
  for (const c of node.children) if (!applyModelLock(c, ctx)) allLocked = false;
  if (allLocked) node.lock_reason = "Required by current settings";
  return allLocked;
}

async function buildModelNodes(ctx: LockContext): Promise<StorageNode[]> {
  const modelsDir = paths().modelsDir;
  const out: StorageNode[] = [];
  let users: AsyncIterable<Deno.DirEntry>;
  try {
    users = Deno.readDir(modelsDir);
  } catch {
    return out;
  }
  for await (const user of users) {
    if (!user.isDirectory) continue;
    const userPath = join(modelsDir, user.name);
    let repos: AsyncIterable<Deno.DirEntry>;
    try {
      repos = Deno.readDir(userPath);
    } catch {
      continue;
    }
    for await (const repo of repos) {
      if (!repo.isDirectory) continue;
      const repoPath = join(userPath, repo.name);
      const display = `${user.name}/${repo.name}`;
      const files: StorageNode[] = [];
      await collectModelFiles(repoPath, repoPath, files);
      files.sort((a, b) => a.name.localeCompare(b.name));
      if (files.length === 0) continue;
      const hasMmproj = files.some((f) => basename(f.name).toLowerCase().startsWith("mmproj"));
      if (hasMmproj || files.length > 1) {
        const size = files.reduce((s, f) => s + f.size, 0);
        out.push(folderNode(display, repoPath, size, files));
      } else {
        const f = files[0];
        out.push(fileNode(`${display}/${f.name}`, f.path, f.size));
      }
    }
  }
  out.sort((a, b) => a.name.localeCompare(b.name));
  for (const n of out) applyModelLock(n, ctx);
  return out;
}

// --- category descriptors --------------------------------------------------

interface CategoryDescriptor {
  id: StorageCategoryId;
  label: string;
  deletable: boolean;
  buildNodes(ctx: LockContext): Promise<StorageNode[]>;
  deleteNode(node: StorageNode): Promise<void>;
}

const CATEGORIES: CategoryDescriptor[] = [
  {
    id: "models",
    label: "Models",
    deletable: true,
    buildNodes: buildModelNodes,
    deleteNode: async (node) => {
      assertUnder(node.path, paths().modelsDir);
      await removePath(node.path);
    },
  },
  {
    id: "binaries",
    label: "Binaries",
    deletable: true,
    async buildNodes(ctx) {
      const list = await binariesManager()
        .list()
        .catch(() => []);
      const out: StorageNode[] = [];
      for (const b of list) {
        if (!b.installed || !b.path) continue;
        const lock = ctx.lockedBinaryKinds.has(b.kind)
          ? "Required by an enabled feature"
          : undefined;
        out.push(fileNode(b.kind, b.path, await statSize(b.path), lock));
      }
      return out;
    },
    deleteNode: async (node) => {
      assertUnder(node.path, paths().binDir);
      await removePath(node.path);
    },
  },
  {
    id: "cache",
    label: "Cache",
    deletable: true,
    async buildNodes() {
      const p = paths();
      const dirs: Array<[string, string]> = [
        ["cache", p.cacheDir],
        ["deno-cache", p.denoCacheDir],
        ["staging", p.stagingDir],
      ];
      const out: StorageNode[] = [];
      for (const [name, dir] of dirs) out.push(folderNode(name, dir, await dirSize(dir)));
      return out;
    },
    deleteNode: async (node) => {
      await removePath(node.path);
      // Cache dirs are long-lived (created by ensureDirs); recreate empty.
      await Deno.mkdir(node.path, { recursive: true });
    },
  },
  {
    id: "logs",
    label: "Logs",
    deletable: true,
    async buildNodes() {
      const p = paths();
      const out: StorageNode[] = [];
      let entries: AsyncIterable<Deno.DirEntry>;
      try {
        entries = Deno.readDir(p.logsDir);
      } catch {
        return out;
      }
      for await (const e of entries) {
        if (!e.isFile) continue;
        const path = join(p.logsDir, e.name);
        // All logs are clearable: the active core.log is truncated (not removed)
        // so the running core keeps logging to it; rotated backups are removed.
        out.push(fileNode(e.name, path, await statSize(path)));
      }
      out.sort((a, b) => a.name.localeCompare(b.name));
      return out;
    },
    deleteNode: async (node) => {
      assertUnder(node.path, paths().logsDir);
      if (normalize(node.path) === normalize(paths().logFile)) {
        // Active log: truncate in place. Deleting the open file wouldn't reclaim
        // space on Unix (the handler keeps the inode) and fails on Windows;
        // truncation empties it now and the logger keeps appending from 0.
        await Deno.truncate(node.path, 0).catch(() => {});
      } else {
        await removePath(node.path);
      }
    },
  },
  {
    id: "sessions",
    label: "Sessions",
    deletable: true,
    buildNodes(ctx) {
      const out = sessionsRepo()
        .listAll()
        .sort((a, b) => b.updatedAtMs - a.updatedAtMs)
        .map((s) =>
          folderNode(
            s.title || "Untitled session",
            sessionDir(s.id),
            s.sizeBytes,
            [],
            ctx.activeSessionIds.has(s.id) ? "Session is active" : undefined,
          ),
        );
      return Promise.resolve(out);
    },
    deleteNode: (node) => {
      sessionsRepo().deleteById(basename(node.path));
      return Promise.resolve();
    },
  },
  {
    id: "toolkits",
    label: "Toolkits",
    deletable: true,
    async buildNodes(ctx) {
      const out: StorageNode[] = [];
      for (const tk of toolkitsRegistry().list()) {
        const dir = toolkitDir(tk.id);
        const lock =
          tk.source === "builtin"
            ? "Built-in toolkit"
            : ctx.enabledToolkitIds.has(tk.id)
              ? "Has enabled tools"
              : undefined;
        out.push(folderNode(tk.displayName || tk.id, dir, await dirSize(dir), [], lock));
      }
      out.sort((a, b) => a.name.localeCompare(b.name));
      return out;
    },
    deleteNode: async (node) => {
      await uninstallToolkit(basename(node.path));
    },
  },
  {
    id: "settings",
    label: "Settings",
    deletable: true,
    async buildNodes() {
      const p = paths();
      const size = (await statSize(p.settingsFile)) + (await statSize(p.secretsEncFile));
      // Never locked; deleting it is a factory reset, not a file removal.
      return [fileNode("Settings + secrets", p.settingsFile, size)];
    },
    deleteNode: async () => {
      await resetCoreSettings();
      await clearAllSecrets();
    },
  },
];

const CATEGORY_BY_ID = new Map(CATEGORIES.map((c) => [c.id, c]));

// --- public API ------------------------------------------------------------

/** Build one category's nodes, degrading to an empty list (logged) on failure
 *  so a single broken category can never blank the whole storage view. */
async function safeBuildNodes(desc: CategoryDescriptor, ctx: LockContext): Promise<StorageNode[]> {
  try {
    return await desc.buildNodes(ctx);
  } catch (err) {
    log.warn(
      `storage category ${desc.id} failed to build: ${err instanceof Error ? err.message : err}`,
    );
    return [];
  }
}

export async function buildStorageTree() {
  const ctx = await buildLockContext();
  const categories: StorageCategory[] = [];
  for (const desc of CATEGORIES) {
    const nodes = await safeBuildNodes(desc, ctx);
    const size = nodes.reduce((s, n) => s + n.size, 0);
    categories.push({ id: desc.id, label: desc.label, deletable: desc.deletable, nodes, size });
  }
  return {
    categories,
    total_size: categories.reduce((s, c) => s + c.size, 0),
    // `<root_path>/models` is the models dir, so the client can still map an HF
    // spec to disk if needed.
    root_path: dirname(paths().modelsDir),
  };
}

/** Delete the given paths. Each must match a node in the freshly-built tree,
 *  belong to a deletable category, and not be locked (nor contain any locked
 *  descendant). The lock is recomputed here server-side, so a stale or tampered
 *  client view can't force a locked delete. Unknown paths are skipped
 *  (idempotent). */
export async function deleteStoragePaths(targets: string[]): Promise<void> {
  const { byPath } = await indexTree();
  const ops: Array<{ desc: CategoryDescriptor; node: StorageNode }> = [];
  for (const t of targets) {
    const hit = byPath.get(normalize(t));
    if (!hit) continue; // already gone
    const { desc, node } = hit;
    if (!desc.deletable) {
      throw new AppError("validation_error", `category ${desc.id} is not deletable`);
    }
    if (subtreeLockReason(node)) {
      throw new AppError("validation_error", `cannot delete in-use item: ${node.path}`);
    }
    ops.push({ desc, node });
  }
  for (const { desc, node } of ops) {
    try {
      await desc.deleteNode(node);
    } catch (err) {
      log.warn(`delete ${node.path} failed: ${err instanceof Error ? err.message : err}`);
      throw err;
    }
  }
  void notifyRequirementsChanged();
}

/** Clear an entire category: delete every non-locked top-level node in it. For
 *  `settings` this performs the factory reset. */
export async function clearStorageCategory(categoryId: string): Promise<void> {
  const desc = CATEGORY_BY_ID.get(categoryId as StorageCategoryId);
  if (!desc) throw new AppError("validation_error", `unknown storage category: ${categoryId}`);
  if (!desc.deletable) {
    throw new AppError("validation_error", `category ${desc.id} is not deletable`);
  }
  const ctx = await buildLockContext();
  const nodes = await safeBuildNodes(desc, ctx);
  for (const node of nodes) {
    if (subtreeLockReason(node)) continue;
    try {
      await desc.deleteNode(node);
    } catch (err) {
      log.warn(`clear ${node.path} failed: ${err instanceof Error ? err.message : err}`);
    }
  }
  void notifyRequirementsChanged();
}

// --- internals -------------------------------------------------------------

async function indexTree(): Promise<{
  byPath: Map<string, { desc: CategoryDescriptor; node: StorageNode }>;
}> {
  const ctx = await buildLockContext();
  const byPath = new Map<string, { desc: CategoryDescriptor; node: StorageNode }>();
  const add = (desc: CategoryDescriptor, node: StorageNode) => {
    byPath.set(normalize(node.path), { desc, node });
    if (node.kind === "folder") for (const c of node.children) add(desc, c);
  };
  for (const desc of CATEGORIES) {
    for (const node of await safeBuildNodes(desc, ctx)) add(desc, node);
  }
  return { byPath };
}

/** The lock reason for a node or, for a folder, the first locked descendant.
 *  Used to refuse deleting anything in use. */
function subtreeLockReason(node: StorageNode): string | undefined {
  if (node.lock_reason) return node.lock_reason;
  if (node.kind === "folder") {
    for (const c of node.children) {
      const r = subtreeLockReason(c);
      if (r) return r;
    }
  }
  return undefined;
}

async function removePath(path: string): Promise<void> {
  try {
    await Deno.remove(path, { recursive: true });
  } catch (err) {
    if (!(err instanceof Deno.errors.NotFound)) throw err;
  }
}

// Reject any path that doesn't sit strictly inside `root` (no `..` escape, no
// absolute re-root, not the root itself).
function assertUnder(target: string, root: string): void {
  const rel = relative(normalize(root), normalize(target));
  if (rel === "" || rel === "." || rel.startsWith("..") || isAbsolute(rel)) {
    throw new AppError("validation_error", `path is not under ${root}: ${target}`);
  }
}
